src/Compilers/Core/MSBuildTask/ManagedCompiler.cs

```diff
@@ -1089,6 +1089,86 @@ private void NormalizePaths(ITaskItem[]? taskItems)
             }
         }
 
+        /// <summary>
+        /// Transforms source file paths to avoid misinterpretation as command-line switches.
+        /// On Unix, source files in the root directory (starting with '/') are transformed
+        /// to start with '/./' to prevent them from being interpreted as switches.
+        /// </summary>
+        /// <param name="sources">The source files to transform</param>
+        /// <returns>Transformed source files suitable for passing to the compiler command line</returns>
+        private ITaskItem[]? GetSourcesForCommandLine(ITaskItem[]? sources)
+        {
+            if (sources is null || sources.Length == 0)
+                return sources;
+
+            // On Windows, no transformation is needed
+            if (Path.DirectorySeparatorChar == '\\')
+                return sources;
+
+            // Check if any source file starts with '/' (Unix root directory)
+            bool needsTransformation = false;
+            foreach (var source in sources)
+            {
+                if (source.ItemSpec.StartsWith("/") && !source.ItemSpec.StartsWith("/./"))
+                {
+                    needsTransformation = true;
+                    break;
+                }
+            }
+
+            if (!needsTransformation)
+                return sources;
```

Comment:
Consider going over the sources just once and lazily initialize `transformedSources` if we determine it's necessary.